### PR TITLE
MMF-3528: Inform users how to achieve deeper, full-project analysis

### PR DESCRIPTION
**Pull request by thahnen**
## Summary

For users to understand the differences between using SonarLint out of the box and in connected mode, views / preference pages and workflows were changed in order to inform the user and remove misconceptions.

## Testing

Testing was done via integration tests and manual testing for corner cases.

### CODE REVIEW

File: its/projects/java/java-simple/src/hello/Hello10.java
Comment (thahnen) [inline its/projects/java/java-simple/src/hello/Hello10.java:1]: When analyzing manually files the limit was upgraded to 10 files at least before the pop-up is showing that it might be slow.

File: its/target-platforms/latest.target
Comment (thahnen) [inline its/target-platforms/latest.target:1]: Eclipse 2023-12 doesn't work with the old snapshots update site anymore. It was also abandoned by Mylin a few years ago.

File: org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfiguratorExtension.java
Comment (thahnen) [inline org.sonarlint.eclipse.cdt/src/org/sonarlint/eclipse/cdt/internal/CProjectConfiguratorExtension.java:1]: Changed alongside the JDT implementation and moved to `SonarLintUtils` to work with the notifications in standalone mode as the core plug-in cannot access the sub-plug-ins.

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java [hunk 1]
```diff
@@ -78,20 +79,30 @@ public abstract class AbstractAnalyzeProjectJob<CONFIG extends AbstractAnalysisC
   private final TriggerType triggerType;
   private final boolean shouldClearReport;
   private final Collection<FileWithDocument> files;
-
+  private EnumSet<Language> unavailableLanguagesReference;
```
Comment (thahnen) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java:82]: Even though Java does pass by values, this is actually a reference. There is no way to denote it this way in Java, that's why I named it this way.

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/utils/SonarLintUtils.java [hunk 2]
```diff
@@ -36,6 +36,17 @@
 import org.sonarsource.sonarlint.core.commons.Language;
 
 public class SonarLintUtils {
+  /**
```
Comment (thahnen) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/utils/SonarLintUtils.java:39]: I thought about enhancing the extension point to have it split for whitelisted standalone and connected mode languages but that doesn't work well if you have Eclipse without CDT for example: There should still be notification shown but due to CDT missing it wouldn't work out well.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/actions/AnalysisJobsScheduler.java [hunk 3]
```diff
@@ -35,57 +41,78 @@
 import org.sonarlint.eclipse.core.resource.ISonarLintFile;
 import org.sonarlint.eclipse.core.resource.ISonarLintProject;
 import org.sonarlint.eclipse.ui.internal.SonarLintProjectDecorator;
+import org.sonarlint.eclipse.ui.internal.popup.LanguageFromConnectedModePopup;
 import org.sonarlint.eclipse.ui.internal.util.PlatformUtils;
+import org.sonarsource.sonarlint.core.commons.Language;
 
 public class AnalysisJobsScheduler {
```
Comment (thahnen) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/actions/AnalysisJobsScheduler.java:48]: This was a bit simplified but still works as expected

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeCommand.java [hunk 4]
```diff
@@ -75,9 +78,20 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
 
   private static void runAnalysisJob(Shell shell, Map<ISonarLintProject, Collection<FileWithDocument>> filesPerProject) {
     var totalFileCount = filesPerProject.values().stream().mapToInt(Collection::size).sum();
-    if (totalFileCount > 1 && !askConfirmation(shell)) {
+    
+    if (!SonarLintGlobalConfiguration.ignoreEnhancedFeatureNotifications()) {
+      MessageDialogUtils.enhancedWithConnectedModeInformation(shell, "Are you working with a CI/CD pipeline?",
+        "Running an analysis with SonarQube / SonarCloud in your pipeline might be better suited for analyzing "
+        + "multiple files or a whole project!");
+    } else if (totalFileCount > 10 && !askConfirmation(shell)) {
```
Comment (thahnen) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeCommand.java:86]: This is the reason for the changes in the `java-simple` IT. Marco was also saying to change the constraint because of the noise.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.ui.internal.util;
+
+import java.util.HashSet;
+
+import org.sonarlint.eclipse.ui.internal.notifications.AbstractNotificationPopup;
+
+/** Utility used for reducing stacked up notifications */
+public class PopupUtils {
```
Comment (thahnen) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java:27]: Was speaking to Marco about that. In the future we should handle it the same way for other pop-ups (the notifications) to lower the noise.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeCommand.java (hunk 4 above)
Comment (henryju) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeCommand.java:86]: This change could have been made in a separate ticket/commit, but fair enough.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  private PopupUtils() {
+    // utility class
+  }
+  
+  private static HashSet<Class<? extends AbstractNotificationPopup>> popupsCurrentlyDisplayed = new HashSet<>();
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java:32]:
```suggestion
  private static Set<Class<? extends AbstractNotificationPopup>> popupsCurrentlyDisplayed = new HashSet<>();
```

**Review by henryju — CHANGES_REQUESTED** [review]
I don't really like the way to pass the list of unsupportedLanguages back to the UI, it makes the code hard to read.
We could maybe reuse the `AnalysisListener` so that the UI can react and display the notification if needed?

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java (hunk 1 above)
Comment (thahnen) [reply org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java:82]: This was removed

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/actions/AnalysisJobsScheduler.java [hunk 7]
```diff
@@ -35,57 +41,78 @@
...
 
   private AnalysisJobsScheduler() {
     // utility class, forbidden constructor
   }
+  
+  public static void scheduleAutoAnalysisIfEnabled(AnalyzeProjectRequest request, boolean unsupportedLanguageCheck) {
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/actions/AnalysisJobsScheduler.java]: Done

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java (hunk 6 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/util/PopupUtils.java]: Done

File: its/src/org/sonarlint/eclipse/its/AbstractSonarLintTest.java [hunk 8]
```diff
@@ -104,8 +104,17 @@ public abstract class AbstractSonarLintTest {
...
   
+  @BeforeClass
+  public static final void setUpBeforeClass() {
+    System.setProperty("sonarlint.internal.ignoreEnhancedFeature", "true");
+    System.setProperty("sonarlint.internal.ignoreMissingFeature", "true");
+  }
+  
   @AfterClass
   public static final void cleanupAfterClass() {
+    System.setProperty("sonarlint.internal.ignoreEnhancedFeature", "");
```
[the remark this answers is not in the document]
Comment (thahnen) [reply its/src/org/sonarlint/eclipse/its/AbstractSonarLintTest.java]: Done

File: its/src/org/sonarlint/eclipse/its/StandaloneAnalysisTest.java [hunk 9]
```diff
@@ -61,11 +63,53 @@
...
+  @Before
+  public void removeEnhancedConnectedModeInformationPerTest() {
+    System.setProperty("sonarlint.internal.ignoreEnhancedFeature", "true");
+    System.setProperty("sonarlint.internal.ignoreMissingFeature", "true");
+  }
+  
+  @Test
+  public void analyzeProjectWithMissingLanguageAnalyzers() {
+    // INFO: This test case should display everything!
+    System.setProperty("sonarlint.internal.ignoreEnhancedFeature", "");
```
[the remark this answers is not in the document]
Comment (thahnen) [reply its/src/org/sonarlint/eclipse/its/StandaloneAnalysisTest.java]: Done

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java [hunk 10]
```diff
@@ -202,6 +213,18 @@ private void runAnalysisAndUpdateMarkers(Map<ISonarLintFile, IDocument> docPerFi
     if (!monitor.isCanceled()) {
       updateMarkers(docPerFiles, issuesPerResource, result, triggerType, monitor);
       updateTelemetry(result, start, issuesPerResource);
+      
+      // Collect all the languages we just analyzed and that are unavailable in standalone mode. This will be re-used
+      // to handle it accordingly in the UI (e.g. display notification to the user).
+      var bindingOpt = SonarLintCorePlugin.getServersManager().resolveBinding(getProject());
+      if (bindingOpt.isEmpty()) {
+        var languages = result.languagePerFile().values().stream().collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AbstractAnalyzeProjectJob.java]: Done

File: .gitignore [hunk 11]
```diff
@@ -18,3 +18,5 @@ org.sonarlint.eclipse.ui/lib/
 # macOS files
 **/.DS_Store
 
+# Integration tests
+!/its/projects/connected/.project
```
[the remark this answers is not in the document]
Comment (thahnen) [reply .gitignore:22]: Not every IT project should have the **.project** file saved, like Maven ones.

**Review by henryju — APPROVED** [review]
LGTM
